### PR TITLE
Leave some header related info for inspection

### PR DESCRIPTION
**Pull request by pepyakin**
This change leaves some header related data in chain after block finalization, notably the block number.

During working on Polkadot it's been discovered that removing this data eagerly leads to `System::block_number()` to return 0 if called in the context of a runtime API based on an in-chain block. 

This PR solves that by leaving in in-chain until the next call to `on_initialize` where it gets overwritten. While I was at it I also changed that for other header related data.

### CODE REVIEW

**Comment by bkchr** [comment]
Can you give more context? From your description, this is done by some runtime api function that skips `on_initialize`?

**Comment by pepyakin** [comment]
Sure,

Yeah, I need to execute some runtime APIs in the context of a certain block. By default Runtime API initialize block for me, specifically:

1. It initializes the environment to the next block.
1. It also runs all the initialization logic. That can be heavy at times, with all those stuff schedules to `on_initialize`

Since I am interested in the specified block, not the next block, and because I need as little overhead as possible for those calls, I use `![skip_initialize_block]` for those runtime APIs.

But in that case, due to the code altered in this PR, all calls to `System::block_number()` returns `0`. This PR fixes that, by leaving this data in state for inspection.

**Comment by bkchr** [comment]
Hmm, what do you think if we change this: https://github.com/paritytech/substrate/blob/master/client/service/src/client/client.rs#L1164-L1170

If we call the runtime at a given block, I would expect that we have the runtime in this state and not some "intermediate" state?

**Comment by pepyakin** [comment]
That may help with №1 (although I am not sure how we would exactly change it without breaking current users or without significant code change), but it won't solve №2 since we will still be calling `on_initalize` and `on_runtime_upgrade` and those can be enormously heavy IIUC.

**Comment by bkchr** [comment]
№2 would be solved by just calling `frame_system::Module::initialize` like offchain worker is doing it for example.

**Comment by bkchr** [comment]
I know that we would change the behavior, however I'm not aware of anything that relied on this. I think we may even broke stuff (just think about the polkadot inherent that is having this special check for session boundaries).

I'm just not that sure we want to keep all this stuff in the block. In the end it increases the block state for every block. :shrug:

**Review by bkchr — APPROVED** [review]

**Comment by pepyakin** [comment]
bot merge

File: frame/system/src/lib.rs [hunk 1]
```diff
@@ -1013,15 +1013,16 @@ impl<T: Config> Module<T> {
...
+	/// resulting header for this block.
 	pub fn finalize() -> T::Header {
 		ExecutionPhase::kill();
 		ExtrinsicCount::kill();
 		AllExtrinsicsLen::kill();
 
-		let number = <Number<T>>::take();
-		let parent_hash = <ParentHash<T>>::take();
-		let mut digest = <Digest<T>>::take();
+		let number = <Number<T>>::get();
```
Comment (andresilva) [inline frame/system/src/lib.rs:1023]: Not sure if it's worth adding a comment here on why these aren't cleared. We have a similar comment at the end of this function.

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
bot merge

**Pull request merged**